### ISSUE TITLE
Explorer filter updates

### ISSUE THREAD

**Issue by lodici** [magarena/magarena#849]
- [ ] update highlight on filter button as criteria is changed for immediate feedback.
- [ ] hover mouse over active filter button to show tooltip displaying filter criteria.
- [ ] add individual filter reset button.
- [ ] add option to sort filter values (eg. Sets by date or name).

**Comment by lodici** [comment]
![untitled](https://cloud.githubusercontent.com/assets/2443519/18710052/076d23d2-7ffb-11e6-9aa8-e28fcefa5065.jpg)

**Comment by lodici** [comment]
A lot of updates but (hopefully) the result is more modular, self-contained and maintainable filters.
![classmagic_1_1ui_1_1explorer_1_1filter_1_1buttons_1_1_filter_panel__inherit__graph](https://cloud.githubusercontent.com/assets/2443519/18710580/e1d86480-7ffd-11e6-8677-d2e038941856.png)

![classmagic_1_1ui_1_1explorer_1_1filter_1_1buttons_1_1_filter_panel__coll__graph](https://cloud.githubusercontent.com/assets/2443519/18710575/df7ad83a-7ffd-11e6-99be-da2d2e502c2d.png)

**Comment by ShawnieBoy** [comment]
It might be worth removing the subtypes we don't support from the filter list, might help make it a bit more manageable (Well, not by much, but every bit helps)

**Comment by lodici** [comment]
If there are subtypes which don't return any cards then it seems reasonable to exclude those from the filter list otherwise it would be inconsistent to arbitrarily remove sub-types just because the cards they return are currently unsupported. Sorting the list would certainly make it easier to use though.

**Comment by ShawnieBoy** [comment]
These are the Planes subtypes which we do not implement, list, or support - We do not list the Planes, Phenomena, Schemes etc. card types from the Planechase and Archenemy sets, these subtypes return no cards.

**Comment by lodici** [comment]
I agree, having filter values (in general) that will never return anything is confusing. Can you just remove the unsupported sub-types from MagicSubType? If not and they are required for gameplay then an EnumSet containing just the valid sub-types for filtering is probably the way to go. Your decision. :grin:

**Comment by ShawnieBoy** [comment]
I've gone for the subset similar to the type filters. We may someday get to multiplayer and different play modes... 😉

**Comment by lodici** [comment]
Just added an update - the list of values returned by ```getFilterValues()``` are the values that are displayed in the filter. When you select a checkbox its position in the list is used by ```isCardValid()``` as the MagicSubType ordinal to check. In short, the list of values used by ```getFilterValues()``` and ```isCardValid()``` must be the same, both is size and order (it was comparing ```MagicSubType.values()``` and ```MagicSubType.FILTER_SUBTYPES```).

**Comment by ShawnieBoy** [comment]
Just noticed that Tribal permanents are displayed as 'Tribal Tribal' in the type column - I don't know if that's somehow related.

**Comment by lodici** [comment]
See https://github.com/magarena/magarena/blob/master/src/magic/ui/cardtable/CardTableModel.java#L174.

**Comment by ShawnieBoy** [comment]
Thanks - The `TYPE_ORDER` already puts Tribal at the beginning of the list of types, the `getLongTypeString` is adding back in the SuperTypes, and for some reason includes Tribal, which isn't a supertype (although it kinda behaves like one). Elite is the only Supertype missing from this list.

**Comment by lodici** [comment]
Can you check whether it affects sorting at all? See https://github.com/magarena/magarena/blob/master/src/magic/model/MagicCardDefinition.java#L974.

**Comment by ShawnieBoy** [comment]
It's still comparing the same two lists, if that's what you mean. However this time there isn't a duplicate of Tribal being both a type and supertype, and so displayed twice. Sorting is as follows, with an artifact example:
Artifact
Legendary Artifact
Snow Artifact
Artifact Creature
Legendary Artifact Creature
Snow Artifact Creature
Artifact Land
Legendary Enchantment Artifact
Enchantment Artifact Creature
Tribal Artifact

**Comment by lodici** [comment]
The sorting of the Type column seems to be fine. My point was that I am not totally sure what https://github.com/magarena/magarena/blob/master/src/magic/model/MagicCardDefinition.java#L974 is actually doing and whether it needs updating now that you have updated getLongTypeString(). It appears to be comparing getLongTypeString() if the getTypeString() does not match which does not make much sense to me.

**Comment by lodici** [comment]
The sorting of the Type column seems fine. I was just wondering whether your change to getLongTypeString() would affect the results of https://github.com/magarena/magarena/blob/master/src/magic/model/MagicCardDefinition.java#L974. That is, when sorting cards by type, if the types match then compare using getLongTypeString() But without delving too deeply, it all seems ok.
